### PR TITLE
[sw, tests] Refactor sram_ctrl scramble test and test utils

### PR DESCRIPTION
**Pull request by silvestrst**
The main purpose of this change is to make the test util more
generic. It also includes some suggestions from the other PRs, such as
avoiding dependencies on top_earlgrey in test utils.

- Renamed test utils API to match accepted style
  `sram_ctrl_testutils_<...>`.

- Testutils are more generic, they operate on the passed in address
  instead of top_earlgrey addresses calculated inside the util.

- Init functions have been removed from test utils.

- Unnecessary data structures have been removed from test utils.

- Test has been simplified to only perform a single check at the start
  of RAM, as opposed to former the start and the end of RAM. If
  required, we can always add the end of RAM check later.

In a nutshell, the test util API was simplified and some of this
complexity was moved to the test instead.

Signed-off-by: Silvestrs Timofejevs <silvestrst@lowrisc.org>

### CODE REVIEW

File: sw/device/lib/testing/sram_ctrl_testutils.h [hunk 1]
```diff
@@ -57,36 +20,25 @@ typedef struct sram_ctrl_data {
...
- * Reads and compares data at the beginning and end of RAM.
+ * Reads data from `address` in SRAM and compares against `expected`.
  *
- * The read data is checked for inequality against the `expected` data.
+ * The data is checked for inequality.
  */
-bool sram_ctrl_read_from_ram_neq(const sram_ctrl_t *sram,
-                                 const sram_ctrl_data_t *expected);
+bool sram_ctrl_testutils_read_check_neq(uintptr_t address,
+                                        const sram_ctrl_data_t *expected);
```
Comment (mcy) [inline sw/device/lib/testing/sram_ctrl_testutils.h:41]: Why is this taken by pointer rather than by value? The ABI will pass it by pointer regardless. Since we're refactoring this, can you make this be taken by value?

File: sw/device/tests/sram_ctrl_scrambled_access_test.c [hunk 2]
```diff
@@ -2,14 +2,31 @@
...
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
 const test_config_t kTestConfig;
 
+static dif_sram_ctrl_t sram_ctrl;
+
+/**
+ * Retention SRAM start address in bytes (inclusive).
+ */
+static const uintptr_t kSramStartAddrInBytes =
```
Comment (mcy) [inline sw/device/tests/sram_ctrl_scrambled_access_test.c:24]: Aren't all addreses "in bytes"? I think you can drop the suffix.

**Review by silvestrst — COMMENTED** [review]
Thanks @mcy for the review!

**Review by timothytrippel — APPROVED** [review]
LGTM, thanks @silvestrst !

File: sw/device/lib/testing/sram_ctrl_testutils.c [hunk 3]
```diff
@@ -10,84 +10,26 @@
...
-sram_ctrl_t sram_ctrl_ret_init(void) { return sram_ctrl_init(&sram_ret); }
-
-/**
- * Writes `data` at the `address` in RAM.
- */
-static void write_to_ram(uintptr_t address, const sram_ctrl_data_t *data) {
+void sram_ctrl_testutils_write(uintptr_t address,
+                               const sram_ctrl_data_t *data) {
   mmio_region_t region = mmio_region_from_addr(address);
   for (size_t index = 0; index < SRAM_CTRL_DATA_NUM_WORDS; ++index) {
```
Comment (sriyerg) [inline sw/device/lib/testing/sram_ctrl_testutils.c:16]: It will be useful to allow the test to decide what size payload to read / write to the SRAM. Suggest making `size` as a field of `sram_ctrl_data_t` (or something similar).

Comment (sriyerg) [reply sw/device/lib/testing/sram_ctrl_testutils.c:16]: We may also want to add OOB checks to catch bad arg errors sooner.

**Review by sriyerg — COMMENTED** [review]
Left 2 comments. Really like the much simplified code!

File: sw/device/lib/testing/sram_ctrl_testutils.c (hunk 3 above)
Comment (silvestrst) [reply sw/device/lib/testing/sram_ctrl_testutils.c:16]:
```

Comment (silvestrst) [reply sw/device/lib/testing/sram_ctrl_testutils.c:16]: > We may also want to add OOB checks to catch bad arg errors sooner.

Are these the checks you would like me to add?

Comment (silvestrst) [reply sw/device/lib/testing/sram_ctrl_testutils.c:16]: I think it makes sense to make both as follow-up PRs and merge this in, as long as you are happy with the rest?

Comment (mundaym) [reply sw/device/lib/testing/sram_ctrl_testutils.c:16]: I think it is OK to address @sriyerg's comments as a follow up PR.

File: sw/device/lib/testing/sram_ctrl_testutils.h (hunk 1 above)
Comment (mcy) [reply sw/device/lib/testing/sram_ctrl_testutils.h:41]: There is; it semantically does not take a pointer, but rather takes it by value. This means it cannot be passed in-line. In general, we prefer to pass things by value when it is the correct semantic.

File: sw/device/tests/sram_ctrl_scrambled_access_test.c [hunk 4]
```diff
@@ -38,16 +58,16 @@ static const sram_ctrl_data_t kRamTestPattern2 = {
...
         "Read-Write first pattern failed");
 
-  // Write second pattern to the start and the end of RAM.
-  sram_ctrl_write_to_ram(sram, &kRamTestPattern2);
-  CHECK(sram_ctrl_read_from_ram_eq(sram, &kRamTestPattern2),
-        "Read-Write second pattern failed");
+  // Write second pattern to the start of SRAM, and read it out.
+  sram_ctrl_testutils_write(kSramStartAddr, &kRamTestPattern2);
+  CHECK(sram_ctrl_testutils_read_check_eq(kSramStartAddr, &kRamTestPattern2),
+        "Read-Write first pattern failed");
```
Comment (doreis-lowrisc) [inline sw/device/tests/sram_ctrl_scrambled_access_test.c:70]: s\first\second\

File: sw/device/lib/testing/sram_ctrl_testutils.h (hunk 1 above)
Comment (silvestrst) [reply sw/device/lib/testing/sram_ctrl_testutils.h:41]: I disagree, logically we want to pass a pointer to the `static const` data defined outside of the function. Passing it in this way, makes the intention less ambiguous.

> This means it cannot be passed in-line.

I don't think that this potential micro optimisation has much value in this particular case.

Comment (mundaym) [reply sw/device/lib/testing/sram_ctrl_testutils.h:41]: Given this is a request to refactor existing code I don't think this PR should be blocked on this discussion even if it does touch the relevant code.

Could you put together a small PR refactoring this code to make it pass by value @silvestrst? Then we can judge the change on its own merits without blocking other work.

Comment (silvestrst) [reply sw/device/lib/testing/sram_ctrl_testutils.h:41]: Sure, can do

File: sw/device/tests/sram_ctrl_scrambled_access_test.c [hunk 5]
```diff
@@ -2,14 +2,31 @@
...
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
 const test_config_t kTestConfig;
 
+static dif_sram_ctrl_t sram_ctrl;
+
+/**
+ * Retention SRAM start address (inclusive).
+ */
+static const uintptr_t kSramStartAddr =
```
Comment (mundaym) [inline sw/device/tests/sram_ctrl_scrambled_access_test.c:24]: nit: can you mention that this is the retention SRAM in this constant somewhere? e.g. `kRetSramStartAddr`.

**Review by mundaym — APPROVED** [review]
LGTM modulo @doreis-lowrisc's comment.

File: sw/device/tests/sram_ctrl_scrambled_access_test.c (hunk 5 above)
Comment (silvestrst) [reply sw/device/tests/sram_ctrl_scrambled_access_test.c]: Done.

File: sw/device/tests/sram_ctrl_scrambled_access_test.c (hunk 4 above)
Comment (silvestrst) [reply sw/device/tests/sram_ctrl_scrambled_access_test.c]: Done.

File: sw/device/lib/testing/sram_ctrl_testutils.h (hunk 1 above)
Comment (silvestrst) [reply sw/device/lib/testing/sram_ctrl_testutils.h:41]: Added to #8546

File: sw/device/lib/testing/sram_ctrl_testutils.c (hunk 3 above)
Comment (silvestrst) [reply sw/device/lib/testing/sram_ctrl_testutils.c:16]: Added to #8546
